### PR TITLE
Refactor/Improve String and Path Handling

### PR DESCRIPTION
**Pull request by LRFLEW**
EDIT: The purpose of this PR has changed since I first made it. Below is the new description, followed by the original description (for the purpose of the discussion).

This PR is intended to improve how strings are handled throughout the project. The primary goal is to make the string manipulations safer, including avoiding possible buffer overflows. Below is the list of changes made in this PR:

* `platform_get_path_separator()` and `platform_get_new_line()` have been removed. They have been replaced by the string macros `PATH_SEPARATOR` and `PLATFORM_NEWLINE` to make maintenance easier with the project.

* All instances of `strcpy`, `strncpy`, `strcat`, and `strncat` have been replaced with their respective calls to `safe_strcpy`, `safe_strcat` and `safe_strcat_path`.

* All instances of `sprintf` and `vsprintf` have been replaced with `snprintf` and `vsnprintf` respectively to avoid overflows (and similar for related functions). A solution that informs the program when an overflow occurs would be preferable, but this is fine for now.

* Most (hopefully all) of the functions that modify a string passed as a parameter also takes a length argument for the string to avoid buffer overflows. The major headers affected are `localization/localization.h`, `platform/platform.h`, and `util/util.h`.

* Calls using `strcat`, `strncat`, `safe_strcat`, and `safe_strcat_path` have been looked at, and replaced with more appropriate calls where applicable. Specifically, concatenations that add a file or folder to a path use `safe_strcat_path`, concatenations that add a file extension use `path_append_extension`, and concatenations to get a path to end with a separator use the new `path_end_with_separator`.

* Calls to `safe_strcpy` (and to a lesser extent, `safe_strcat`) have been checked to make sure that the size parameter isn't being abused. Specifically, some calls to `safe_strcpy` have used the size parameter to truncate strings when copying. This caused erroneous warnings in the console about string truncation. This sort of functionality should be provided by a `safe_strncpy` function (that takes both the length of the input string and the output buffer size) if desired, but this PR was able to avoid needing a `strncpy`-like function by using other techniques.

---

Original Description:

I made this after figuring out the crash on macOS (which I initially fixed in #4485) and realized that having both the macro and function for the path separator was redundant. After noticing that the error I fixed priviously was in other parts of the code, I decided to just replace all the usage of `platform_get_path_separator` with `PATH_SEPARATOR`. I also noticed that `platform_get_new_line` served a similar purpose, so I separated it out into a macro as well. This PR includes the fix in #4485.

### CODE REVIEW

**Comment by LRFLEW** [comment]
While working on transitioning the path handling to use `safe_strcat_path`, I noticed that there was a LOT of possible buffer overflow issues. I'm working on fixing as many of these as I can find at once.

**Comment by janisozaur** [comment]
Yes, our string handling is all over the place. I hope you fix as many as you can.

File: src/platform/windows.c [hunk 1]
```diff
@@ -1126,7 +1117,7 @@ static bool windows_setup_file_association(
 	}
 	// [hRootKey\OpenRCT2.ext\DefaultIcon]
 	wchar_t szIconW[MAX_PATH];
-	wsprintfW(szIconW, L"\"%s\",%d", dllPathW, iconIndex);
+	StringCchPrintfW(szIconW, MAX_PATH, L"\"%s\",%d", dllPathW, iconIndex);
```
Comment (IntelOrca) [inline src/platform/windows.c]: How about you use swprintf_s? Or is that not supported on mingw?

Comment (LRFLEW) [inline src/platform/windows.c]: *facepalm* Didn't realize that was a thing 😜. It does seem that it's slightly different, though, as it handles truncation rather differently (though probably in a more preferable way). Don't know if it's in MinGW, though, so I'll have to try a build and see.

Comment (IntelOrca) [inline src/platform/windows.c]: As long as the function always adds the null terminator, even if the string is longer the the buffer size then its fine.

File: src/windows/track_place.c [hunk 2]
```diff
@@ -167,10 +167,6 @@ void window_track_place_open(const track_design_file_ref *tdFileRef)
 	_window_track_place_last_x = 0xFFFF;
 	_currentTrackPieceDirection = (2 - get_current_rotation()) & 3;
 	window_track_place_draw_mini_preview(td6);
-
-	// TODO: 3155 appears to be empty. What is this supposed to do?
-	char *title = (char*)language_get_string(3155);
```
Comment (duncanspumpkin) [inline src/windows/track_place.c]: Why are you removing this?

File: src/windows/track_place.c [hunk 3]
```diff
@@ -445,6 +441,7 @@ static void window_track_place_attempt_placement(rct_track_td6 *td6, int x, int
  */
 static void window_track_place_paint(rct_window *w, rct_drawpixelinfo *dpi)
 {
+	set_format_arg(0, char *, _trackDesign->name);
```
Comment (duncanspumpkin) [inline src/windows/track_place.c]: I don't think this is a very clever way of doing this. I much prefer the older way of assigning it to 3155.

File: src/windows/track_manage.c [hunk 4]
```diff
@@ -240,6 +236,7 @@ static void window_track_manage_invalidate(rct_window *w)
  */
 static void window_track_manage_paint(rct_window *w, rct_drawpixelinfo *dpi)
 {
+	set_format_arg(0, char *, _trackDesignFileReference->name);
```
Comment (duncanspumpkin) [inline src/windows/track_manage.c]: You've forgotten to change the 3155 here.

File: src/windows/track_place.c (hunk 3 above)
Comment (IntelOrca) [inline src/windows/track_place.c]: @duncanspumpkin no we can't do that anymore!

File: src/windows/track_manage.c (hunk 4 above)
Comment (LRFLEW) [inline src/windows/track_manage.c]: Fixed, thanks for pointing that out.

File: src/platform/windows.c (hunk 1 above)
Comment (LRFLEW) [inline src/platform/windows.c]: > even if the string is longer the the buffer size

I'm not sure that's particularly ok 😜. Either way, the different handling is that, when the resultant string is longer than the buffer, the buffer is set to an empty string (and maybe an error number is returned).

Comment (IntelOrca) [inline src/platform/windows.c]: @LRFLEW I am talking about the fact that strncpy does not add a null terminator if it has to truncate the string.

**Comment by LRFLEW** [comment]
This latest commit should (hopefully) force all of the code in `localization.c` to avoid overflow on any input buffers (except for maybe `win1252_to_utf8`, which I haven't gotten to yet). It changes the behavior of certain things, which might break some spots I haven't seen yet (in particular, the `_part` functions now set `*dest` to point to the null character, not the character after it). Overflow can be checked for in the `_part` functions by seeing if `*size` (which is generally the length of the string excluding the null character) was set to 0 (which is the exception; if `*size` is 0, the string was truncated, but was also null-terminated). I want to add a warning when truncation happens (like with the `safe_` functions), but I haven't added that yet. Builds will fail because I haven't pushed the other related changes yet (I wanted to make sure this change went under enough review).

**Comment by LRFLEW** [comment]
With the exception of OS-dependent code (I've been testing on macOS, and using the preprocessor to help), I have removed calls to a number of insecure functions:

All calls to `strcpy`, `strncpy`, `strcat`, and `strncat` have been removed. They have been replaced with `safe_strcpy`, `safe_strcat` and `safe_strcat_path` where appropriate. (NOTE: There are probably still a few cases in the code where `safe_strcat` is called where `safe_strcat_path` would be more appropriate. It shouldn't affect the result, as it would only be in places where the string to be concatenated already ends in the path eliminator, but it would be nice to fix those cases for semantic reasons).

`sprintf` is a tricky situation. Much like the other functions, the version that Microsoft considers safe is not always available on the other platforms. We could implement a `safe_sprintf`, but I don't think re-implementing the `printf` core functionality is something we want to do, and any cheats we could use would be system-dependent. For now, I've changed to code to use `snprintf`, which *is* safe in terms of always producing the null character (Microsoft considers it unsafe due to how the format parameter is validated, IIRC). I think the best plan for `sprintf` and `snprintf`, though, is to replace all the functionality with string concatenation and the localization functions (which I have made safe in a previous commit).

**Comment by LRFLEW** [comment]
Since this has basically become me making the code more compliant with Microsoft's "Security Development Lifecycle" (SDL) checks, I'll drop a link here from Microsoft that I'm using to check what functions are banned with SDL checks and what their recommended replacement functions are.

https://msdn.microsoft.com/en-us/library/bb288454.aspx

I realized that many of the functions say *Recommended*, but I'm not certain how those work with the SDL checks. (maybe they throw warnings instead of errors; I don't know). `snprintf` is a "*Recommended*" banned function, so maybe it's not as high of a priority. Similarly, `strlen` is a "*Recommended*" banned function. Oddly enough, the only function that I know is still in the code (after this PR) that's is banned (and not *Recommended*) is `memcpy`. However, the replacement doesn't *seem* that much safer, IMO. (The difference is that the safe version (`memcpy_s`) take a length for both the input and the output). I don't know if that's something we want tackled in this PR or not.

**Comment by janisozaur** [comment]
C11 has `memcpy_s` as part of standard

**Comment by LRFLEW** [comment]
But not C++11, or C++14. I haven't seen anything about C++17 having it either.

**Comment by LRFLEW** [comment]
Also, worth pointing out that while `memcpy` is considered unsafe with SDL, [`std::copy_n`](http://en.cppreference.com/w/cpp/algorithm/copy_n) is fine.

**Comment by LRFLEW** [comment]
1000+ lines of code changes later, and I think this is done. I'll edit the first message to better explain what this PR has changed. I believe I've got everything that needs to be changed. Feel free to point out anything I've missed.

**Pull request merged**